File: mobile/lib/loan.js
// Loan and amortization math. Pure functions, no network. The whole point of
// this file is to show a Filipino borrower the TRUE cost of a loan, and to stop
// a lender's low "add-on" rate from hiding an effective rate roughly double it.
//
// Two rate conventions matter here:
//   Diminishing balance: interest each month is charged on the REMAINING
//     balance. This is how banks and formal lenders amortize. The standard
//     amortization formula applies.
//   Add-on: interest is charged on the ORIGINAL principal for the whole term,
//     then spread evenly. Common in in-house and informal financing. It looks
//     cheap but the effective rate is far higher, because you keep paying
//     interest on money you have already paid back.
//
// This is an estimate tool. Real contracts add fees, penalties, and
// pre-termination charges the UI must disclose. No em or en dashes in copy.

const num = (x) => (Number.isFinite(Number(x)) ? Number(x) : 0);
const round2 = (x) => Math.round(num(x) * 100) / 100;

// Level monthly payment on a diminishing-balance loan.
//   A = P * r / (1 - (1 + r)^-n), and A = P / n when the rate is zero.
export function monthlyPayment(principal, monthlyRate, months) {
  const P = Math.max(0, num(principal));
  const r = Math.max(0, num(monthlyRate));
  const n = Math.max(1, Math.round(num(months)));
  if (P === 0) return 0;
  if (r === 0) return round2(P / n);
  const factor = r / (1 - Math.pow(1 + r, -n));
  return round2(P * factor);
}

// Full diminishing-balance schedule. Each row splits the payment into interest
// on the current balance and principal, and the balance reconciles to zero. The
// last payment absorbs any rounding so the loan closes exactly.
export function amortize(principal, monthlyRate, months) {
  const P = Math.max(0, num(principal));
  const r = Math.max(0, num(monthlyRate));
  const n = Math.max(1, Math.round(num(months)));
  const payment = monthlyPayment(P, r, n);
  const schedule = [];
  let balance = P;
  let totalInterest = 0;
  for (let i = 1; i <= n; i++) {
    const interest = round2(balance * r);
    let principalPaid = round2(payment - interest);
    // The final row clears whatever is left, absorbing rounding drift.
    if (i === n || principalPaid > balance) principalPaid = round2(balance);
    const rowPayment = round2(principalPaid + interest);
    balance = round2(balance - principalPaid);
    totalInterest = round2(totalInterest + interest);
    schedule.push({ period: i, payment: rowPayment, interest, principal: principalPaid, balance: Math.max(0, balance) });
  }
  return {
    payment,
    months: n,
    totalInterest: round2(totalInterest),
    totalPaid: round2(P + totalInterest),
    schedule,
  };
}

// Add-on loan: interest = principal * monthly add-on rate * months, charged on
// the original principal for the whole term. Payment is the level split of
// principal plus that interest.
export function addOnLoan(principal, monthlyAddOnRate, months) {
  const P = Math.max(0, num(principal));
  const rate = Math.max(0, num(monthlyAddOnRate));
  const n = Math.max(1, Math.round(num(months)));
  const totalInterest = round2(P * rate * n);
  const totalPaid = round2(P + totalInterest);
  const payment = round2(totalPaid / n);
  return { payment, months: n, totalInterest, totalPaid };
}

// The effective monthly rate a loan really costs, backed out from its principal,
// its level payment, and its term. For an add-on loan this reveals the true
// rate hiding behind the quoted one. Solved by bisection since the present-value
// function is monotonic in the rate.
export function effectiveMonthlyRate(principal, payment, months) {
  const P = Math.max(0, num(principal));
  const A = Math.max(0, num(payment));
  const n = Math.max(1, Math.round(num(months)));
  if (P <= 0 || A <= 0) return 0;
  if (A * n <= P) return 0; // paid no more than borrowed, so no interest
  // Present value of n level payments of A at monthly rate r.
  const pv = (r) => (r === 0 ? A * n : A * (1 - Math.pow(1 + r, -n)) / r);
  let lo = 0;
  let hi = 1; // 100% a month is a generous upper bound for any real loan
  for (let i = 0; i < 200; i++) {
    const mid = (lo + hi) / 2;
    // pv decreases as r rises; if pv is still above P, the rate must be higher.
    if (pv(mid) > P) lo = mid;
    else hi = mid;
  }
  return (lo + hi) / 2;
}

// Effective annual rate from a monthly rate, with compounding.
export function effectiveAnnualRate(monthlyRate) {
  const r = Math.max(0, num(monthlyRate));
  return Math.pow(1 + r, 12) - 1;
}

// loanSummary(principal, ratePercent, months, opts) -> one object the screen can
// render directly. opts.method 'diminishing' (default) or 'addon';
// opts.rateBasis 'monthly' (default) or 'annual' for how the rate is quoted.
// Returns the payment, totals, the schedule, and BOTH the nominal and the true
// effective annual rate so the real cost is never hidden.
export function loanSummary(principal, ratePercent, months, opts = {}) {
  const P = Math.max(0, num(principal));
  const n = Math.max(1, Math.round(num(months)));
  const method = opts && opts.method === 'addon' ? 'addon' : 'diminishing';
  const rateBasis = opts && opts.rateBasis === 'annual' ? 'annual' : 'monthly';
  // The quoted rate as a monthly decimal.
  const quotedMonthly = rateBasis === 'annual' ? num(ratePercent) / 100 / 12 : num(ratePercent) / 100;

  let payment, totalInterest, totalPaid, schedule;
  if (method === 'addon') {
    const a = addOnLoan(P, quotedMonthly, n);
    payment = a.payment;
    totalInterest = a.totalInterest;
    totalPaid = a.totalPaid;
    // Rebuild a diminishing schedule at the loan's TRUE rate so the row split is
    // honest, then let the last row absorb rounding to the same total.
    const eff = effectiveMonthlyRate(P, payment, n);
    schedule = amortize(P, eff, n).schedule;
  } else {
    const am = amortize(P, quotedMonthly, n);
    payment = am.payment;
    totalInterest = am.totalInterest;
    totalPaid = am.totalPaid;
    schedule = am.schedule;
  }

  const effMonthly = effectiveMonthlyRate(P, payment, n);
  const effAnnual = effectiveAnnualRate(effMonthly);

  return {
    principal: P,
    months: n,
    method,
    payment,
    totalInterest,
    totalPaid,
    quotedMonthlyRate: round4(quotedMonthly),
    effectiveMonthlyRate: round4(effMonthly),
    effectiveAnnualRate: round4(effAnnual),
    schedule,
  };
}

// Interest saved by paying the loan off after `paidMonths` payments, on a
// diminishing-balance loan (the only kind where early payoff cuts interest).
export function payoffSaving(principal, monthlyRate, months, paidMonths) {
  const full = amortize(principal, monthlyRate, months);
  const k = Math.max(0, Math.min(Math.round(num(paidMonths)), full.months));
  if (k >= full.months) return { interestSaved: 0, balanceCleared: 0 };
  const interestPaidSoFar = full.schedule.slice(0, k).reduce((s, row) => s + row.interest, 0);
  const balanceCleared = k > 0 ? full.schedule[k - 1].balance : num(principal);
  const interestSaved = round2(full.totalInterest - interestPaidSoFar);
  return { interestSaved, balanceCleared: round2(balanceCleared) };
}

function round4(x) {
  return Math.round(num(x) * 10000) / 10000;
}

File: flutter/lib/money/loan.dart
// Loan and amortization math, ported 1:1 from mobile/lib/loan.js. The whole
// point is to show a Filipino borrower the TRUE cost of a loan, and to stop
// a lender's low add-on rate from hiding an effective rate roughly double
// it. Two conventions: diminishing balance (banks; interest on the
// remaining balance) and add-on (in-house and informal; interest on the
// original principal for the whole term, which looks cheap and is not).
// This is an estimate tool; real contracts add fees the UI must disclose.
// Golden-verified against the real RN module.

import 'dart:math' as math;

import 'ledger.dart' show amountOf;

double _jsRound(num x) => (x + 0.5).floorToDouble();

double _round2(dynamic x) => _jsRound(amountOf(x) * 100) / 100;

double _round4(dynamic x) => _jsRound(amountOf(x) * 10000) / 10000;

/// A hard ceiling on the term, so the schedule can never grow without
/// bound and freeze the UI.
const int maxMonths = 1200;

int _clampMonths(dynamic months) {
  final rounded = _jsRound(amountOf(months));
  final atLeastOne = rounded > 1 ? rounded : 1.0;
  return (atLeastOne < maxMonths ? atLeastOne : maxMonths.toDouble()).toInt();
}

/// Level monthly payment on a diminishing-balance loan.
/// A = P * r / (1 - (1 + r)^-n), and A = P / n when the rate is zero.
double monthlyPayment(dynamic principal, dynamic monthlyRate, dynamic months) {
  final p = math.max(0.0, amountOf(principal));
  final r = math.max(0.0, amountOf(monthlyRate));
  final n = _clampMonths(months);
  if (p == 0) return 0;
  if (r == 0) return _round2(p / n);
  final factor = r / (1 - math.pow(1 + r, -n));
  return _round2(p * factor);
}

/// Full diminishing-balance schedule. Each row splits the payment into
/// interest on the current balance and principal; the last payment absorbs
/// rounding so the loan closes exactly.
Map<String, dynamic> amortize(
    dynamic principal, dynamic monthlyRate, dynamic months) {
  final p = math.max(0.0, amountOf(principal));
  final r = math.max(0.0, amountOf(monthlyRate));
  final n = _clampMonths(months);
  final payment = monthlyPayment(p, r, n);
  final schedule = <Map<String, dynamic>>[];
  var balance = p;
  var totalInterest = 0.0;
  for (var i = 1; i <= n; i++) {
    final interest = _round2(balance * r);
    var principalPaid = _round2(payment - interest);
    if (i == n || principalPaid > balance) principalPaid = _round2(balance);
    final rowPayment = _round2(principalPaid + interest);
    balance = _round2(balance - principalPaid);
    totalInterest = _round2(totalInterest + interest);
    schedule.add({
      'period': i,
      'payment': rowPayment,
      'interest': interest,
      'principal': principalPaid,
      'balance': balance > 0 ? balance : 0.0,
    });
  }
  return {
    'payment': payment,
    'months': n,
    'totalInterest': _round2(totalInterest),
    'totalPaid': _round2(p + totalInterest),
    'schedule': schedule,
  };
}

/// Add-on loan: interest charged on the ORIGINAL principal for the whole
/// term, then spread evenly.
Map<String, dynamic> addOnLoan(
    dynamic principal, dynamic monthlyAddOnRate, dynamic months) {
  final p = math.max(0.0, amountOf(principal));
  final rate = math.max(0.0, amountOf(monthlyAddOnRate));
  final n = _clampMonths(months);
  final totalInterest = _round2(p * rate * n);
  final totalPaid = _round2(p + totalInterest);
  final payment = _round2(totalPaid / n);
  return {
    'payment': payment,
    'months': n,
    'totalInterest': totalInterest,
    'totalPaid': totalPaid,
  };
}

/// The effective monthly rate a loan really costs, backed out from its
/// principal, level payment, and term by bisection. For an add-on loan
/// this reveals the true rate hiding behind the quoted one.
double effectiveMonthlyRate(dynamic principal, dynamic payment, dynamic months) {
  final p = math.max(0.0, amountOf(principal));
  final a = math.max(0.0, amountOf(payment));
  final n = _clampMonths(months);
  if (p <= 0 || a <= 0) return 0;
  if (a * n <= p) return 0;
  double pv(double r) =>
      r == 0 ? a * n : a * (1 - math.pow(1 + r, -n)) / r;
  var lo = 0.0;
  var hi = 1.0;
  while (pv(hi) > p && hi < 1e6) {
    hi *= 2;
  }
  for (var i = 0; i < 200; i++) {
    final mid = (lo + hi) / 2;
    if (pv(mid) > p) {
      lo = mid;
    } else {
      hi = mid;
    }
  }
  return (lo + hi) / 2;
}

/// Effective annual rate from a monthly rate, with compounding.
double effectiveAnnualRate(dynamic monthlyRate) {
  final r = math.max(0.0, amountOf(monthlyRate));
  return math.pow(1 + r, 12) - 1.0;
}

/// One object the screen renders directly. method 'diminishing' (default)
/// or 'addon'; rateBasis 'monthly' (default) or 'annual'. Returns the
/// payment, totals, the schedule, and BOTH the nominal and the true
/// effective annual rate so the real cost is never hidden.
Map<String, dynamic> loanSummary(
    dynamic principal, dynamic ratePercent, dynamic months,
    {String method = 'diminishing', String rateBasis = 'monthly'}) {
  final p = math.max(0.0, amountOf(principal));
  final n = _clampMonths(months);
  final chosenMethod = method == 'addon' ? 'addon' : 'diminishing';
  final chosenBasis = rateBasis == 'annual' ? 'annual' : 'monthly';
  final quotedMonthly = chosenBasis == 'annual'
      ? amountOf(ratePercent) / 100 / 12
      : amountOf(ratePercent) / 100;

  double payment, totalInterest, totalPaid;
  List<Map<String, dynamic>> schedule;
  if (chosenMethod == 'addon') {
    final a = addOnLoan(p, quotedMonthly, n);
    payment = a['payment'] as double;
    totalInterest = a['totalInterest'] as double;
    totalPaid = a['totalPaid'] as double;
    // Rebuild a diminishing schedule at the loan's TRUE rate so the row
    // split is honest; the last row absorbs the rounding drift so the
    // interest column reconciles to the add-on headline.
    final eff = effectiveMonthlyRate(p, payment, n);
    schedule = (amortize(p, eff, n)['schedule'] as List)
        .cast<Map<String, dynamic>>();
    final sumInterest = _round2(
        schedule.fold(0.0, (s, row) => s + (row['interest'] as double)));
    final drift = _round2(totalInterest - sumInterest);
    if (schedule.isNotEmpty && drift != 0) {
      final last = schedule.last;
      last['interest'] = _round2((last['interest'] as double) + drift);
      last['payment'] = _round2((last['payment'] as double) + drift);
    }
  } else {
    final am = amortize(p, quotedMonthly, n);
    payment = am['payment'] as double;
    totalInterest = am['totalInterest'] as double;
    totalPaid = am['totalPaid'] as double;
    schedule = (am['schedule'] as List).cast<Map<String, dynamic>>();
  }

  final effMonthly = effectiveMonthlyRate(p, payment, n);
  final effAnnual = effectiveAnnualRate(effMonthly);

  return {
    'principal': p,
    'months': n,
    'method': chosenMethod,
    'payment': payment,
    'totalInterest': totalInterest,
    'totalPaid': totalPaid,
    'quotedMonthlyRate': _round4(quotedMonthly),
    'nominalAnnualRate': _round4(quotedMonthly * 12),
    'effectiveMonthlyRate': _round4(effMonthly),
    'effectiveAnnualRate': _round4(effAnnual),
    'schedule': schedule,
  };
}

/// Interest saved by paying off after `paidMonths` payments, on a
/// diminishing-balance loan (the only kind where early payoff cuts
/// interest).
Map<String, dynamic> payoffSaving(
    dynamic principal, dynamic monthlyRate, dynamic months, dynamic paidMonths) {
  final full = amortize(principal, monthlyRate, months);
  final fullMonths = full['months'] as int;
  final rounded = _jsRound(amountOf(paidMonths));
  final k = math.max(0, math.min(rounded.toInt(), fullMonths));
  if (k >= fullMonths) return {'interestSaved': 0.0, 'balanceCleared': 0.0};
  final schedule = (full['schedule'] as List).cast<Map<String, dynamic>>();
  var interestPaidSoFar = 0.0;
  for (final row in schedule.take(k)) {
    interestPaidSoFar += row['interest'] as double;
  }
  final balanceCleared =
      k > 0 ? schedule[k - 1]['balance'] as double : amountOf(principal);
  return {
    'interestSaved': _round2((full['totalInterest'] as double) - interestPaidSoFar),
    'balanceCleared': _round2(balanceCleared),
  };
}
